experiment/gerrit-onboarder/onboarder.go

```diff
@@ -341,73 +345,134 @@ func ensureProjectConfig(config, host, cur_branch, groupName string) (string, er
 		return fmt.Errorf("failed to switch to new branch, %w", err)
 	}
 
 	return nil
 }
 
+func execInDir(stdout, stderr io.Writer, dir string, cmd string, args ...string) error {
+	(&logrus.Logger{
+		Out:       os.Stderr,
+		Formatter: logrus.StandardLogger().Formatter,
+		Hooks:     logrus.StandardLogger().Hooks,
+		Level:     logrus.StandardLogger().Level,
+	}).WithField("dir", dir).
+		WithField("cmd", cmd).
+		// The default formatting uses a space as separator, which is hard to read if an arg contains a space
+		WithField("args", fmt.Sprintf("['%s']", strings.Join(args, "', '"))).
+		Info("running command")
+
+	c := exec.Command(cmd, args...)
+	c.Dir = dir
+	c.Stdout = stdout
+	c.Stderr = stderr
+	return c.Run()
+}
+
+func createCR(workDir string) error {
+	diff, err := getDiff(workDir)
+	if err != nil {
+		return err
+	}
+	if diff == "" {
+		logrus.Info("No changes made. Returning without creating CR")
+		return nil
+	}
+	commitMessage := fmt.Sprintf("Grant the Prow cluster read and label permissions\n\nChange-Id: I%s", bumper.GitHash(fmt.Sprintf("%d", rand.Int())))
+	if err := execInDir(os.Stdout, os.Stderr, workDir, "git", "commit", "-a", "-v", "-m", commitMessage); err != nil {
+		return fmt.Errorf("unable to commit: %w", err)
+	}
+	if err := execInDir(os.Stdout, os.Stderr, workDir, "git", "push", "origin", "HEAD:refs/for/refs/meta/config"); err != nil {
+		return fmt.Errorf("unable to push: %w", err)
+	}
+	return nil
+}
+
+func getDiff(workDir string) (string, error) {
+	var diffBuf bytes.Buffer
+	var errBuf bytes.Buffer
+	if err := execInDir(&diffBuf, &errBuf, workDir, "git", "diff"); err != nil {
+		return "", fmt.Errorf("diffing previous bump: %v -- %s", err, errBuf.String())
+	}
+	return diffBuf.String(), nil
+}
+
 func main() {
 	o, err := parseAndValidateOptions()
 	if err != nil {
 		logrus.Fatal(err)
 	}
 
+	workDir, err := ioutil.TempDir("", "gerrit_onboarding")
+	if err != nil {
+		logrus.Fatal(err)
+	}
+	defer os.RemoveAll(workDir)
+
+	if err := execInDir(os.Stdout, os.Stderr, workDir, "git", "clone", fmt.Sprintf("sso://%s/%s", o.host, o.repo)); err != nil {
+		logrus.Fatal(fmt.Errorf("failed to clone sso://%s/%s %w", o.host, o.repo, err))
+	}
+
+	workDir = path.Join(workDir, o.repo)
+
 	rand.Seed(time.Now().UTC().UnixNano())
 	branchName := fmt.Sprintf("gerritOnboarding_%d", rand.Int())
 
-	if err = fetchMetaConfig(o.host, o.repo, branchName); err != nil {
+	if err = fetchMetaConfig(o.host, o.repo, branchName, workDir); err != nil {
 		logrus.Fatal(err)
 	}
 
-	groupsChanged, err := updateGroups(o.uuID, o.groupName)
+	groupsChanged, err := updateGroups(workDir, o.uuID, o.groupName)
 	if err != nil {
 		logrus.Fatal(err)
 	}
 
 	if groupsChanged {
 		logrus.Info("groups file did not change. Skipping stash step")
-		if err := bumper.Call(os.Stdout, os.Stderr, "git", "stash"); err != nil {
+		if err := execInDir(os.Stdout, os.Stderr, workDir, "git", "stash"); err != nil {
```

Comment:
This doesn't match the log. Also, I assume `git stash` is because of checking out different repos in the same dir, should we consider checking out in different dir to save one shelling out to git?

Reply:
Yea. The stash _was_ necessary to traverse the inheritance tree when checking the project config, but I realized that if we do that first we never need to stash changes. That is much cleaner.